components/server/src/workspace/workspace-service.ts

```diff
@@ -561,6 +561,12 @@ export class WorkspaceService {
                 daysToLive = daysToLive * 2;
             }
             deletionEligibilityTime.setDate(deletionEligibilityTime.getDate() + daysToLive);
+            if (new Date() > deletionEligibilityTime) {
```

Comment:
```suggestion
            if (workspace.deletionEligibilityTime > deletionEligibilityTime.toISOString()) {
```

Let's not get too magic with automatic conversions :wink:

Reply:
@geropl sounds good, but your suggestion uses `workspace.deletionEligibilityTime` instead of the current time? I thought we agreed we would use the current time as a simpler check to avoid the git status update problem where we would not be able to shorten the lifespan from 28 days to 14 if there were no more uncommitted/unpushed changes?

Reply:
>  I thought we agreed we would use the current time as a simpler check

I didn't even notice we were discussing this...? :sweat_smile: What I meant with "prevent set it to the "past"" was "don't move the pointer backwards", e.g. comparing current and new `deletionEligibilityTime`, which is the stricter and IMO "correct'er" invariant.

Reply:
@geropl I get not wanting to move the pointer backwards, but also think it has a big shortcoming in a pretty probable scenario:
- The user starts a workspace, works on the code (there are unpushed commits, untracked files, etc.) and therefore, the `deletionEligibilityTime` is now 28 days in the future (the standard 14 * 2 because the workspace is dirty).
- Now the user commits everything, pushes to the remote and `updateGitStatus` is called again. This new `deletionEligibilityTime` update would be ignored because the new proposed time would be today + 14 days.

Your proposed change would make it so that the de facto only garbage collection happens after 28 days, and only be at 14 for workspaces that were never touched. Maybe I'm missing something, but that's my current understanding of how the git status updates work.

Reply:
> Your proposed change would make it so that the de facto only garbage collection happens after 28 days, and only be at 14 for workspaces that were never touched.

@filiptronicek Exactly. Maybe we have to differentiate our goals: I'm first and foremost interested in understanding what the root cause to our current behavior is. That's why I'm interested in logging every time we are "moving the pointer back". Whether we keep some workspaces for longer is worth it IMO, and we can turn it back.
In general I assume we need multiple attempts here.

That's also the reason why I would suggest we keep the DB query the same: that way we ensure we surface this behavior again if it happens regularly.
I wouldn't feel too good about fixing the problem here, without understanding why it surfaced in the first place. :confused: